lib/RestructureCFG/SimplifyDualSwitch.cpp

```diff
@@ -42,9 +42,26 @@ computeSwitchToIfPromotion(SwitchNode *Switch) {
   size_t CasesNum = Switch->cases_size();
   Case ElectedCase = Invalid;
 
-  // We cannot promote switches with more than 2 cases
-  if (Switch->cases_size() > 2) {
-    return std::nullopt;
+  // The promotion logic varies depending on the fact that we are treating a
+  // dispatcher `switch` or a standard `switch`
+  ASTNode *DefaultCase = Switch->getDefault();
+  if (Switch->getCondition() == nullptr) {
+    revng_assert(DefaultCase == nullptr);
+
+    // We cannot promote switches with more than 2 cases. No default case is
+    // present on dispatcher switches by construction
+    if (Switch->cases_size() > 2) {
+      return std::nullopt;
+    }
+  } else {
+
+    // We cannot promote switches with more than 2 cases, or switches with two
+    // cases that do not have a default case (the promotion would be
+    // semantically incorrect)
+    size_t CasesNum = Switch->cases_size();
+    if (CasesNum > 2 or (CasesNum == 2 and DefaultCase == nullptr)) {
+      return std::nullopt;
+    }
```

Comment:
This is correct, strictly speaking. But it's also much more convoluted than it needs to be.

I think the following would be more concise and always correct.

```c++
// We cannot promote switches with more than 2 cases
size_t CaseNum = Switch->cases_size();
if (CaseNum > 2)
  return std::nullopt;

// We cannot promote non-dispatcher switches with exactly 2 cases and without
// default, because it would be semantically incorrect, given that the implicit
// default for a switch without a default is to jump past the switch.
if (Switch->getCondition() != nullptr and CaseNum == 2 and Switch->getDefault() == nullptr)
  return std::nullopt;
```

Reply:
Done in 78abb1a4cb266e6932a0462c535feb83756b385e